packages/cli/src/util/update-notifier/index.ts

```diff
@@ -0,0 +1,103 @@
+import semver from 'semver';
+import XDGAppPaths from 'xdg-app-paths';
+import { dirname, resolve as resolvePath } from 'path';
+// import { fileURLToPath } from 'url';
+import { outputJSON, readJSON } from 'fs-extra';
+import type { PackageJson } from '@vercel/build-utils';
+import { spawn } from 'child_process';
+
+interface UpdateNotifierConfig {
+  version: string;
+  expireAt: number;
+  notified: boolean;
+}
+
+/**
+ * Determines if it needs to check for a newer CLI version and returns the last
+ * detected version. The version could be stale, but still newer than the
+ * current version.
+ */
+export default async function updateNotifier({
+  cacheDir = XDGAppPaths('com.vercel.cli').cache(),
+  distTag = 'latest',
+  pkg,
+  updateCheckInterval = 1000 * 60 * 60 * 24 * 7, // 1 week
+  wait,
+}: {
+  cacheDir?: string;
+  distTag?: string;
+  pkg: PackageJson;
+  updateCheckInterval?: number;
+  wait?: boolean;
+}): Promise<string | undefined> {
+  const cacheFile = resolvePath(
+    cacheDir,
+    'update-notifier',
+    `${pkg.name}-${distTag}.json`
+  );
+
+  const loadCache = async (): Promise<UpdateNotifierConfig | undefined> => {
+    try {
+      return await readJSON(cacheFile);
+    } catch (e: any) {
+      // cache does not exist or malformed
+    }
+  };
+
+  let cache = await loadCache();
+
+  if (!cache || cache.expireAt < Date.now()) {
+    await new Promise<void>(resolve => {
+      // spawn the worker, wait for the worker to report it's ready, then
+      // signal the worker to fetch the latest version
+
+      const script = resolvePath(
+        dirname(__filename),
+        '..',
+        '..',
+        '..',
+        'dist',
+        'update-worker.js'
+      );
+      // const script = resolvePath(dirname(fileURLToPath(import.meta.url)), '..', '..', '..', 'dist', 'update-worker.js');
```

Comment:
The commented out line was simply to help @TooTallNate with the ESM conversion. Once we move to ESM, we no longer have `__filename` and I thought it would be more helpful to have the line ready to go than have to waste cycles porting the current `script` value.

Reply:
I removed this line not because it was unused, but because I found an issue in resolving the path. It will need to be eventually be updated to use `fileURLToPath(import.meta.url)` when the CLI moves to ESM.